### DISCUSSION TITLE
Enforce policies with hooks / middleware

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#1877] [Ideas]
As mentioned by @kelseyhightower :) A very successful feature of Kubernetes is the ability to extend it with custom policies out-of-band from actual configurations. So an admin can inject a policy once, and everything will have to run through it.

There are 2 parallel tracks available to implement this:

1. Cue itself is well suited for policy injection. I defer to @mpvl @myitcv @verdverm for details and examples.
2. Runtime hooks. This is what Kubernetes does. It could be webhooks; or it could be executable hooks running inside dagger.

Details TBD :) This is a placeholder so we don’t forget this important feature.

**Comment by verdverm** [comment]
@mpvl has a nice write up here: https://github.com/cuelang/cue/discussions/818 **Using CUE for policy definition and admission control**

**Comment by kinthaiofficial** [comment]
Enforcing policies with hooks/middleware in Dagger pipelines is an important enterprise feature — you want guardrails that apply consistently across all pipelines without requiring each pipeline author to implement them.

**The use cases:**
- Security: no container images from untrusted registries
- Compliance: all builds must produce a SBOM, all containers must be signed
- Cost: pipelines above a certain complexity require manager approval
- Quality: no deployments to production without passing specific test stages

**The middleware model:**
The cleanest implementation is a "pipeline wrapper" pattern — a shared library that wraps pipeline functions with policy checks:

```python
# In a shared Dagger module
@dagger.function
async def with_policy_enforcement(
    pipeline: dagger.Callable,
    context: PolicyContext,
) -> dagger.Container:
    # Pre-pipeline checks
    await check_image_registry_policy(context)
    await check_required_approvals(context)
    
    # Run actual pipeline
    result = await pipeline()
    
    # Post-pipeline checks
    await require_sbom(result)
    await require_signature(result)
    
    return result
```

**The hook lifecycle:**
- `before_build`: validate inputs, check approvals, set up audit logging
- `after_build`: validate outputs, generate SBOMs, sign artifacts
- `on_failure`: emit alerts, create incident tickets

**Policy-as-code:**
Store policies in a versioned repository. The middleware reads policies from the repo rather than hardcoding them. Policy changes are PRs with reviews, just like code changes.

**Enforcement vs observability:**
Start with observability (log policy violations without blocking). Build confidence in the policy rules. Then switch to enforcement (fail the pipeline on policy violation). Running "warn only" for a sprint before enforcing prevents surprises.
